File: backend/src/main/kotlin/com/plugga/backend/entity/Deck.java
package com.plugga.backend.entity;

import com.fasterxml.jackson.annotation.JsonIgnore;

import javax.persistence.*;
import java.sql.Timestamp;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "deck")
public class Deck {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "deck_id")
    private int id;

    @Column(name = "name")
    private String name;

    @Column(name = "image_url")
    private String imageUrl;

    @Column(name = "date_created")
    private Timestamp dateCreated;

    @OneToMany(mappedBy = "deck",
            cascade = {CascadeType.DETACH, CascadeType.MERGE, CascadeType.PERSIST, CascadeType.REFRESH})
    private List<DeckCard> cards;

    @OneToMany(mappedBy = "deck")
    private List<UserDeck> users;

    public Deck() {
    }

    public Deck(final int id) {
        this.id = id;
        this.name = "";
        this.imageUrl = "";
    }

    public Deck(String name, String imageUrl, Timestamp dateCreated) {
        this.name = name;
        this.imageUrl = imageUrl;
        this.dateCreated = dateCreated;
    }

    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getImageUrl() {
        return imageUrl;
    }

    public void setImageUrl(String imageUrl) {
        this.imageUrl = imageUrl;
    }

    public Timestamp getDateCreated() {
        return dateCreated;
    }

    public void setDateCreated(Timestamp dateCreated) {
        this.dateCreated = dateCreated;
    }

    @JsonIgnore
    public List<DeckCard> getCards() {
        return cards;
    }

    public void setCards(final List<DeckCard> cards) {
        this.cards = cards;
    }

    @JsonIgnore
    public List<UserDeck> getUsers() {
        return users;
    }

    public void setUsers(final List<UserDeck> users) {
        this.users = users;
    }

    @Override
    public String toString() {
        return "Deck{" +
                "id=" + id +
                ", name='" + name + '\'' +
                ", imageUrl='" + imageUrl + '\'' +
                ", dateCreated=" + dateCreated +
                '}';
    }

    public void addCard(DeckCard card) {
        if (cards == null) {
            cards = new ArrayList<>();
        }
        cards.add(card);
    }
}

File: backend/src/main/kotlin/com/plugga/backend/entity/Deck.kt
package com.plugga.backend.entity

import com.fasterxml.jackson.annotation.JsonIdentityInfo
import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.ObjectIdGenerators
import com.plugga.backend.jackson.EntityIdResolver
import java.sql.Timestamp
import javax.persistence.Column
import javax.persistence.Entity
import javax.persistence.GeneratedValue
import javax.persistence.GenerationType
import javax.persistence.Id
import javax.persistence.OneToMany
import javax.persistence.Table

@Entity
@Table(name = "deck")
@JsonIdentityInfo(
        generator = ObjectIdGenerators.PropertyGenerator::class,
        property = "id",
        resolver = EntityIdResolver::class,
        scope = Deck::class
)
class Deck {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "deck_id")
    var id: Int = 0

    @Column(name = "name")
    var name: String? = null

    @Column(name = "image_url")
    var imageUrl: String? = null

    @Column(name = "date_created", updatable = false)
    var dateCreated: Timestamp? = null

    @OneToMany(mappedBy = "deck")
    @get:JsonIgnore
    var cards: MutableList<DeckCard>? = null

    @OneToMany(mappedBy = "deck")
    @get:JsonIgnore
    var users: MutableList<UserDeck>? = null

    @OneToMany(mappedBy = "deck")
    @get:JsonIgnore
    var piles: MutableList<Pile>? = null

    constructor() {}

    constructor(id: Int) {
        this.id = id
        this.name = ""
        this.imageUrl = ""
    }

    constructor(name: String, imageUrl: String, dateCreated: Timestamp) {
        this.name = name
        this.imageUrl = imageUrl
        this.dateCreated = dateCreated
    }

    override fun toString(): String {
        return "Deck{" +
                "id=" + id +
                ", name='" + name + '\''.toString() +
                ", imageUrl='" + imageUrl + '\''.toString() +
                ", dateCreated=" + dateCreated +
                '}'.toString()
    }

    fun addCard(card: DeckCard) {
        if (cards == null) {
            cards = mutableListOf()
        }
        cards!!.add(card)
    }
}
